### PR TITLE
Core-edge network improvements

### PR DESCRIPTION
**Pull request by davidegrohmann**

### CODE REVIEW

File: enterprise/core-edge/src/main/java/org/neo4j/coreedge/catchup/CatchupServerProtocol.java [hunk 1]
```diff
@@ -28,9 +28,9 @@ public void expect( NextMessage nextMessage )
         this.nextMessage = nextMessage;
     }
 
-    public boolean isExpecting( NextMessage message )
+    NextMessage expecting()
```
Comment (martinfurmanski) [inline enterprise/core-edge/src/main/java/org/neo4j/coreedge/catchup/CatchupServerProtocol.java]: I think isExpecting was clearer really, but no big deal.

File: enterprise/core-edge/src/main/java/org/neo4j/coreedge/catchup/RequestDecoderDispatcher.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    }
+
+    @Override
+    public void channelRead( ChannelHandlerContext ctx, Object msg ) throws Exception
+    {
+        NextMessage expecting = protocol.expecting();
+        ChannelInboundHandler delegate = decoders.get( expecting );
+        if ( delegate == null )
+        {
+            log.warn( "Unknown message %s", expecting );
```
Comment (martinfurmanski) [inline enterprise/core-edge/src/main/java/org/neo4j/coreedge/catchup/RequestDecoderDispatcher.java]: Maybe "Unknown message, was expecting %s". Perhaps even print a few bytes.

File: enterprise/core-edge/src/main/java/org/neo4j/coreedge/catchup/RequestDecoderDispatcher.java [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+import io.netty.channel.ChannelInboundHandlerAdapter;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import org.neo4j.coreedge.catchup.CatchupServerProtocol.NextMessage;
+import org.neo4j.logging.Log;
+import org.neo4j.logging.LogProvider;
+
+public class RequestDecoderDispatcher extends ChannelInboundHandlerAdapter
```
Comment (martinfurmanski) [inline enterprise/core-edge/src/main/java/org/neo4j/coreedge/catchup/RequestDecoderDispatcher.java]: Could this extend MessageToMessageDecoder<ByteBuf> like the other decoders?

File: enterprise/core-edge/src/main/java/org/neo4j/coreedge/catchup/ServerMessageTypeHandler.java [hunk 4]
```diff
@@ -43,7 +43,7 @@ public ServerMessageTypeHandler( CatchupServerProtocol protocol, LogProvider log
     @Override
     public void channelRead( ChannelHandlerContext ctx, Object msg ) throws Exception
     {
-        if ( protocol.isExpecting( NextMessage.MESSAGE_TYPE ) )
+        if ( protocol.expecting().equals( NextMessage.MESSAGE_TYPE ) )
```
Comment (martinfurmanski) [inline enterprise/core-edge/src/main/java/org/neo4j/coreedge/catchup/ServerMessageTypeHandler.java]: NextMessage is such a stupid class name.

File: enterprise/core-edge/src/test/java/org/neo4j/coreedge/catchup/RequestDecoderDispatcherTest.java [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+    private final AssertableLogProvider logProvider = new AssertableLogProvider();
+
+    @Test
+    public void shouldDispatchToRegisteredDecoder() throws Exception
+    {
+        // given
+        RequestDecoderDispatcher dispatcher = new RequestDecoderDispatcher( protocol, logProvider );
+        protocol.expect( TX_PULL );
+        ChannelInboundHandler delegate = mock( ChannelInboundHandler.class );
+        dispatcher.register( TX_PULL, delegate );
```
Comment (martinfurmanski) [inline enterprise/core-edge/src/test/java/org/neo4j/coreedge/catchup/RequestDecoderDispatcherTest.java]: Could you register at least 3 of them and check that the 2nd one is called?

File: enterprise/core-edge/src/main/java/org/neo4j/coreedge/catchup/RequestDecoderDispatcher.java (hunk 3 above)
Comment (davidegrohmann) [inline enterprise/core-edge/src/main/java/org/neo4j/coreedge/catchup/RequestDecoderDispatcher.java]: `MessageToMessageDecoder` does a lot of thing behind the scene which are not needed for our delegation mechanism here.

**Comment by davidegrohmann** [comment]
@martinfurmanski I have fixed your comments if you would like to have another look at it

**Pull request merged**